File: algo_2021/w5.-traverse-divide-and-conquer/596.Minimum_Subtree/596_minimum_subtree__gb.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: the root of binary tree
    @return: the root of the minimum subtree
    """
    def __init__(self):
        self.minimum_weight = float('inf')
        self.minimum_subtree_root = None
    
    def findSubtree(self, root):
        # write your code here
        self.get_tree_sum(root)
        return self.minimum_subtree_root
    
    def get_tree_sum(self, root):
        if root is None:
            return 0
        
        left_weight = self.get_tree_sum(root.left)
        right_weight = self.get_tree_sum(root.right)
        root_weight = left_weight + right_weight + root.val
        
        if root_weight < self.minimum_weight:
            self.minimum_subtree_root = root
            self.minimum_weight = root_weight
        return root_weight

File: lintcode_algo_2021/s5.-traverse-d-and-c/596.Minimum_Subtree/596_minimum_subtree__gb.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: the root of binary tree
     * @return: the root of the minimum subtree
     */
    private int minSum;
    private TreeNode minRoot;
    public Solution() {
       minSum = Integer.MAX_VALUE;
       minRoot = null; 
    }
    public TreeNode findSubtree(TreeNode root) {
        // write your code here
        getSum(root);
        
        return minRoot;
    }

    private int getSum(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int leftWeight = getSum(root.left);
        int rightWeight = getSum(root.right);
        int rootWeight = leftWeight + rightWeight + root.val;

        if (rootWeight < minSum) {
            minSum = rootWeight;
            minRoot = root;
        }
        return rootWeight;
    }
}